Drafts folder poluted with tons of drafts

### Expected behavior
When writing an email, Mail should create one draft and update it regularly as I type it.

### Actual behavior
When writing an email, Mail creates a new draft every few seconds:

![image](https://user-images.githubusercontent.com/8179031/98085302-68a01900-1e7d-11eb-8201-bbc425bbe966.png)

### Mail app

**Mail app version:** 1.5.1

I can reproduce this issue on master. Another finding: When I open the outdated drafts they don't load and I get the following output in the browser console:

```
[DEBUG] mail: the  draft ID changed, we have to fetch the draft {app: "mail", uid: "admin", currentId: "no draft currently loaded", newId: 78}
xhr.js:184 GET http://localhost/index.php/apps/mail/api/messages/78/body 404 (Not Found)
ConsoleLogger.js:54 [ERROR] mail: could not load draft 78 {error: TypeError: Cannot read property 'databaseId' of undefined}
``` 
Only the most recent on can be displayed and contains the correct text. It seems like the old drafts are deleted from the backend but stay in the vue store.

--- a/lib/Listener/DeleteDraftListener.php
+++ b/lib/Listener/DeleteDraftListener.php
@@ -32,11 +32,13 @@ use OCA\Mail\Db\Mailbox;
 use OCA\Mail\Db\MailboxMapper;
 use OCA\Mail\Db\Message;
 use OCA\Mail\Events\DraftSavedEvent;
+use OCA\Mail\Events\MessageDeletedEvent;
 use OCA\Mail\Events\MessageSentEvent;
 use OCA\Mail\IMAP\IMAPClientFactory;
 use OCA\Mail\IMAP\MessageMapper;
 use OCP\AppFramework\Db\DoesNotExistException;
 use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventDispatcher;
 use OCP\EventDispatcher\IEventListener;
 use Psr\Log\LoggerInterface;
 
@@ -54,14 +56,19 @@ class DeleteDraftListener implements IEventListener {
 	/** @var LoggerInterface */
 	private $logger;
 
+	/** @var IEventDispatcher */
+	private $eventDispatcher;
+
 	public function __construct(IMAPClientFactory $imapClientFactory,
 								MailboxMapper $mailboxMapper,
 								MessageMapper $messageMapper,
-								LoggerInterface $logger) {
+								LoggerInterface $logger,
+								IEventDispatcher $eventDispatcher) {
 		$this->imapClientFactory = $imapClientFactory;
 		$this->mailboxMapper = $mailboxMapper;
 		$this->messageMapper = $messageMapper;
 		$this->logger = $logger;
+		$this->eventDispatcher = $eventDispatcher;
 	}
 
 	public function handle(Event $event): void {
@@ -105,6 +112,10 @@ class DeleteDraftListener implements IEventListener {
 				'exception' => $e,
 			]);
 		}
+
+		$this->eventDispatcher->dispatchTyped(
+			new MessageDeletedEvent($account, $draftsMailbox, $draft->getUid())
+		);
 	}
 
 	/**

--- a/src/components/NewMessageDetail.vue
+++ b/src/components/NewMessageDetail.vue
@@ -53,6 +53,7 @@ export default {
 			originalBody: undefined,
 			errorMessage: '',
 			error: undefined,
+			newDraftId: undefined,
 		}
 	},
 	computed: {
@@ -156,7 +157,12 @@ export default {
 			// `saveDraft` replaced the current URL with the updated draft UID
 			// in that case we don't really start a new draft but just keep the
 			// URL consistent, hence not loading anything
-			if (to.name === 'message' && this.draft && to.params.draftId === parseInt(this.draft.databaseId, 10)) {
+			if (to.name === 'message' && this.draft
+				&& (
+					to.params.draftId === parseInt(this.draft.databaseId, 10)
+					|| to.params.draftId === this.newDraftId
+				)
+			) {
 				logger.debug('detected navigation to current (new) draft UID, not reloading')
 				return
 			}
@@ -282,15 +288,41 @@ export default {
 				body: data.isHtml ? data.body.value : toPlain(data.body).value,
 			}
 			const { id } = await saveDraft(data.account, dataForServer)
+
+			// Remove old draft envelope
+			this.$store.commit('removeEnvelope', { id: data.draftId })
+			this.$store.commit('removeMessage', { id: data.draftId })
+
+			// Fetch new draft envelope
+			await this.$store.dispatch('fetchEnvelope', id)
+
+			// Update route to new draft (actual redirect will be skipped)
+			const account = this.$store.getters.getAccount(data.account)
+			if (parseInt(this.$route.params.mailboxId, 10) === account.draftsMailboxId) {
+				this.newDraftId = id
+				this.$router.replace({
+					to: 'message',
+					params: {
+						mailboxId: this.$route.params.mailboxId,
+						threadId: this.$route.params.threadId,
+						draftId: id,
+					},
+				})
+			}
+
 			return id
 		},
-		sendMessage(data) {
+		async sendMessage(data) {
 			logger.debug('sending message', { data })
 			const dataForServer = {
 				...data,
 				body: data.isHtml ? data.body.value : toPlain(data.body).value,
 			}
-			return sendMessage(data.account, dataForServer)
+			await sendMessage(data.account, dataForServer)
+
+			// Remove old draft envelope
+			this.$store.commit('removeEnvelope', { id: data.draftId })
+			this.$store.commit('removeMessage', { id: data.draftId })
 		},
 	},
 }

--- a/tests/Unit/Listener/DeleteDraftListenerTest.php
+++ b/tests/Unit/Listener/DeleteDraftListenerTest.php
@@ -32,6 +32,7 @@ use OCA\Mail\Db\Mailbox;
 use OCA\Mail\Db\MailboxMapper;
 use OCA\Mail\Db\Message;
 use OCA\Mail\Events\DraftSavedEvent;
+use OCA\Mail\Events\MessageDeletedEvent;
 use OCA\Mail\Events\MessageSentEvent;
 use OCA\Mail\IMAP\IMAPClientFactory;
 use OCA\Mail\IMAP\MessageMapper;
@@ -41,6 +42,7 @@ use OCA\Mail\Model\NewMessageData;
 use OCA\Mail\Model\RepliedMessageData;
 use OCP\AppFramework\Db\DoesNotExistException;
 use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventDispatcher;
 use OCP\EventDispatcher\IEventListener;
 use PHPUnit\Framework\MockObject\MockObject;
 use Psr\Log\LoggerInterface;
@@ -59,6 +61,9 @@ class DeleteDraftListenerTest extends TestCase {
 	/** @var LoggerInterface|MockObject */
 	private $logger;
 
+	/** @var IEventDispatcher|MockObject */
+	private $eventDispatcher;
+
 	/** @var IEventListener */
 	private $listener;
 
@@ -69,12 +74,14 @@ class DeleteDraftListenerTest extends TestCase {
 		$this->mailboxMapper = $this->createMock(MailboxMapper::class);
 		$this->messageMapper = $this->createMock(MessageMapper::class);
 		$this->logger = $this->createMock(LoggerInterface::class);
+		$this->eventDispatcher = $this->createMock(IEventDispatcher::class);
 
 		$this->listener = new DeleteDraftListener(
 			$this->imapClientFactory,
 			$this->mailboxMapper,
 			$this->messageMapper,
-			$this->logger
+			$this->logger,
+			$this->eventDispatcher
 		);
 	}
 
@@ -100,6 +107,8 @@ class DeleteDraftListenerTest extends TestCase {
 			->method('addFlag');
 		$this->logger->expects($this->never())
 			->method('error');
+		$this->eventDispatcher->expects($this->never())
+			->method('dispatchTyped');
 
 		$this->listener->handle($event);
 	}
@@ -181,6 +190,8 @@ class DeleteDraftListenerTest extends TestCase {
 			->method('addFlag');
 		$this->logger->expects($this->never())
 			->method('error');
+		$this->eventDispatcher->expects($this->never())
+			->method('dispatchTyped');
 
 		$this->listener->handle($event);
 	}
@@ -235,6 +246,15 @@ class DeleteDraftListenerTest extends TestCase {
 			->with('Drafts');
 		$this->logger->expects($this->never())
 			->method('error');
+
+		$messageDeletedEvent = new MessageDeletedEvent(
+			$account,
+			$mailbox,
+			$draft->getUid()
+		);
+		$this->eventDispatcher->expects($this->once())
+			->method('dispatchTyped')
+			->with($messageDeletedEvent);
 
 		$this->listener->handle($event);
 	}

